components/vr/webvr_thread.rs

```diff
@@ -0,0 +1,349 @@
+                    self.handle_get_displays(sender);
+                    self.schedule_poll_events();
+                },
+                WebVRMsg::GetFrameData(pipeline_id, device_id, near, far, sender) => {
+                    self.handle_framedata(pipeline_id, device_id, near, far, sender);
+                },
+                WebVRMsg::ResetPose(pipeline_id, device_id, sender) => {
+                    self.handle_reset_pose(pipeline_id, device_id, sender);
+                },
+                WebVRMsg::RequestPresent(pipeline_id, device_id, sender) => {
+                    self.handle_request_present(pipeline_id, device_id, sender);
+                },
+                WebVRMsg::ExitPresent(pipeline_id, device_id, sender) => {
+                    self.handle_exit_present(pipeline_id, device_id, sender);
+                },
+                WebVRMsg::CreateCompositor(device_id) => {
+                    self.handle_create_compositor(device_id);
+                },
+                WebVRMsg::Exit => {
+                    break
+                },
+            }
+        }
+    }
+
+    fn handle_register_context(&mut self, ctx: PipelineId) {
+        self.contexts.insert(ctx);
+    }
+
+    fn handle_unregister_context(&mut self, ctx: PipelineId) {
+        self.contexts.remove(&ctx);
+    }
+
+    fn handle_get_displays(&mut self, sender: IpcSender<WebVRResult<Vec<VRDisplayData>>>) {
+        let devices = self.service.get_devices();
+        let mut displays = Vec::new();
+        for device in devices {
+            displays.push(device.borrow().display_data());
+        }
+        sender.send(Ok(displays)).unwrap();
+    }
+
+    fn handle_framedata(&mut self,
+                        pipeline: PipelineId,
+                        device_id: u64,
+                        near: f64,
+                        far: f64,
+                        sender: IpcSender<WebVRResult<VRFrameData>>) {
+      match self.access_check(pipeline, device_id) {
+            Ok(device) => {
+                sender.send(Ok(device.borrow().inmediate_frame_data(near, far))).unwrap()
+            },
+            Err(msg) => sender.send(Err(msg.into())).unwrap()
+        }
+    }
+
+    fn handle_reset_pose(&mut self,
+                         pipeline: PipelineId,
+                         device_id: u64,
+                         sender: IpcSender<WebVRResult<VRDisplayData>>) {
+        match self.access_check(pipeline, device_id) {
+            Ok(device) => {
+                device.borrow_mut().reset_pose();
+                sender.send(Ok(device.borrow().display_data())).unwrap();
+            },
+            Err(msg) => {
+                sender.send(Err(msg.into())).unwrap()
+            }
+        }
+    }
+
+    fn access_check(&self, pipeline: PipelineId, device_id: u64) -> Result<&VRDevicePtr, &'static str> {
+        if *self.presenting.get(&device_id).unwrap_or(&pipeline) != pipeline {
+            return Err("Device owned by another context");
```

Comment:
`Device` -> `VR Display`

`owned by another context` -> can we be more descriptive?